Replace textarea with tui-textarea

pm100 opened the pull request:
PR for #1662 

now using tui-textarea v.04, that adds keyboard based text selection (hold down shift key while moving cursor with mouse)

reminder that a newline is made by pressing ctrl-enter or shift-enter

pm100 commented:
still would like to point out that one of the tests fails on windows due to a privilege issue. It would be nice if that test was cfg !windows

extrawurst commented on src/components/textinput.rs:
```diff
@@ -1,3 +1,5 @@
+#![allow(unused_imports)]
```
lets not

extrawurst commented on src/components/textinput.rs:
```diff
@@ -14,39 +16,105 @@ use crate::{
 	Password,
 }
+/*
+	completely rewritten using tui-textarea as it provides a ton of useful features
```
lets write comments as-is not as-compared-to-before. comments on the diff belong into the PR desciption

extrawurst commented on src/components/textinput.rs:
```diff
@@ -14,39 +16,105 @@ use crate::{
+
+All keys work as before.
+
+New line is shift-enter, ctrl-enter . These are all common new line editor commands (discord, emacs,...)
```
neither of the two work for me on mac-os. lets make them work & show up in the command list

extrawurst commented on src/components/textinput.rs:
```diff
@@ -14,39 +16,105 @@ use crate::{
+New line is shift-enter, ctrl-enter . These are all common new line editor commands (discord, emacs,...)
+
+There is no help for the editor window. The only thing a user really needs to know is the newline key stroke,
+but there are 10-15 ctrl key codes. I could add then to the general help popup, or make a special one for textinput
```
I like the idea of a specific help popup for the editor

extrawurst commented on src/components/textinput.rs:
```diff
@@ -14,39 +16,105 @@ use crate::{
 	embed: bool,
-	char_count: usize,
+	textarea: Option<TextAreaComponent>,
+	selecting: bool,
```
these two selection specific ones probably make sense to be an enum or struct and might fix the struct_excessive_bools-warning.

extrawurst commented on src/components/textinput.rs:
```diff
@@ -57,20 +125,23 @@ impl TextInputComponent {
 		show_char_count: bool,
 	) -> Self {
+		// this is here purely to make clippy happy
+		let eat_white_space = strings::symbol::WHITESPACE;
```
if this PR removes needing `strings::symbol::WHITESPACE` then lets remove it instead of this stunt

pm100 replied:
OK, I got fixated on keeping the changes to one file

extrawurst commented on src/components/textinput.rs:
```diff
@@ -84,14 +155,30 @@ impl TextInputComponent {
-		self.msg.clear();
-		self.update_count();
-		self.cursor_position = 0;
+		self.msg.take(); // = String::new();
```
comment not needed

extrawurst commented on src/components/textinput.rs:
```diff
@@ -84,14 +155,30 @@ impl TextInputComponent {
 
 	/// Get the `msg`.
+
+	// the fancy footwork with the OnceCell is to allow
```
lets put the comment (non doc comment) inside the fn

extrawurst commented on src/components/textinput.rs:
```diff
@@ -109,120 +196,54 @@ impl TextInputComponent {
-		self.msg = msg;
-		self.cursor_position = 0;
-		self.update_count();
+		self.msg.take();
```
instead of `take`+`set` we can just do `self.msg = msg.into();`, right?

pm100 replied:
I thought there had to be a way to do it in one call.

extrawurst reviewed (changes requested):
a few nitpicks but ultimately the primary problem is: for me in none of the macos terminals (iterm2, Terminal, vscode embedded) the newline command works

pm100 commented:
Its odd because I partially developed on mac so this shift-enter must be new-ish problem. Anyway there is a crossterm issue that I have reported (https://github.com/crossterm-rs/crossterm/issues/861). It is immediately resolvable by changing the startup code of gitui where it sets up crossterm. But that only works in some terminals (ITerm2 works, apple terminal does not).

I can also put back the ctrl-M code for newline that was there way back but one early reviewer said they didnt like it so I took it out.

I will wait a few days to see what crossterm has to say

pm100 commented:
temp close while trying out fixes

pm100 commented:
I have added ctrl-m back , meaing 'new line'

I have added the code to partially fix the mac issue, in my tests it works on iterm2, but not plain terminal. I dont think thats a good way to do it since its now not 100% clear to the user what will work when. The code is there so you can verify that it works.

Fixed everything else (pedantic is a pain, I think some things are less clear now, but whatever..)

Regarding adding help popup , extra key help etc... Can we get this big change in first. Everything works as before so nobody looses anything by not knowing the keystrokes

extrawurst commented:
@pm100 I went back through https://github.com/extrawurst/gitui/pull/509 which contains most of the considerations around this new-line command issue. lets go back to that original plan and use a new binding like `ctr+o` to commit and raw `enter` for the newline.

pm100 commented:
I think thats a really bad idea, people will hate it

- its a breaking change
- its a breaking change to the most often used path through gitui ('gitui', 'a','c','xxxxx', 'enter')
- its a breaking change that is introduced solely to enable a feature that will not be used 99% of the time (multi line commit messages)
- its not obvious that ctrl-o means commit. It feels like 'open' to me
- its a breaking change caused by working around a bug in crossterm that only affects mac users using the native terminal

If you use my suggestion (ctrl-m for new line) then the feature is enabled, and nothing breaks, its an obscure key combination for use in a rare case. And maybe disable ctrl-enter on terminals where it doesnt work. 

Its your codebase but I guarantee you will annoy lots of users. Including me, because my muscle memory will still hit enter every time (just like I hit Esc every time to close a popup dialog because that seems natural - and it does close many popups, Annoys me every time it happens). I will be annoyed but at least I know the solution, others will not know what to do.

extrawurst commented:
i am unsure:

its been a long discussion prior. see old PR. 

its not a stable version of gitui yet, it will be documented in the cmd-bar. if anything can be broken its before 1.0.

Plus: your muscle memory will lead to no breaking change, you will insert a new-line, newcomers might be put of by the fact right now that enter is not a new-line and does a commit.

plus plus: you can re-bind the command if you want to.

pm100 commented:
How about this
- I make it so that you can specify the key binding for new line (not doable at the moment )
-  I make it so there is only one binding for newline (there are multiple ones now, and your key config does not support that)
- I add to the bottom line of the text input box text saying something like 'ctrl-n=newline'
- I make the default ctrl-n (I dont like ctrl-M - its a default from tui-textarea but I override all those anyway, I think it comes from emacs)
- I think the commit = enter keystroke is also configurable, if not I will make it so.

Making it hard to do the no brainer thing is a really bad idea. But with the above changes people who want to make it hard to do the 99% case and easy to do the 1% case can do exactly that. Everybody happy.

pm100 commented:
I have modified the PR so that all the changes I suggest are there, just so you can test it out and see

Here is the key config file for people who like it the hard way :-)
```
(
  multiline_enter: Some((code:Char('o') , modifiers:"CONTROL")),
  textbox_newline: Some((code:Enter, modifiers:"NONE")),
)
```

Includes fixing the textinput clients that should be single line (like branch name etc)

extrawurst commented on src/components/create_branch.rs:
```diff
@@ -104,7 +104,8 @@ impl CreateBranchComponent {
 				true,
-			),
+			)
+			.with_input_type(InputType::Singleline),
```
hm good catch, this was already an oversight before but now its getting to be a very small input field. let me see if I can fix that on master separately

extrawurst replied:
<img width="612" alt="Screenshot 2024-02-15 at 21 00 51" src="https://github.com/extrawurst/gitui/assets/776816/46925689-c02e-40da-968e-22122c397738">

extrawurst commented on src/keys/key_list.rs:
```diff
@@ -209,6 +211,8 @@ impl Default for KeysList {
 			update_submodule: GituiKeyEvent::new(KeyCode::Char('u'),  KeyModifiers::empty()),
 			commit_history_next: GituiKeyEvent::new(KeyCode::Char('n'),  KeyModifiers::CONTROL),
+			multiline_enter: GituiKeyEvent::new(KeyCode::Enter,  KeyModifiers::empty()),
+			textbox_newline: GituiKeyEvent::new(KeyCode::Char('r'),  KeyModifiers::CONTROL),
```
lets call it just `newline`

extrawurst commented on src/main.rs:
```diff
@@ -282,13 +286,29 @@ fn run_app(
 fn shutdown_terminal() {
 	let leave_screen =
-		io::stdout().execute(LeaveAlternateScreen).map(|_f| ());
+		stdout().execute(LeaveAlternateScreen).map(|_f| ());
```
lets leave it consistent with above, this change is not needed

extrawurst commented on src/main.rs:
```diff
@@ -299,6 +319,15 @@ fn shutdown_terminal() {
+		Ok(true)
+	);
+	if supports_keyboard_enhancement {
+		crossterm::queue!(io::stdout(), PopKeyboardEnhancementFlags)
```
i dont know yet how to reproduce this but I get this printed after closing gitui using this branch:

```
PopKeyboardEnhancementFlags failed%
```

how can we get rid of this?

pm100 replied:
since we are not using ctrl/shift enter anymore I will remove all these changes, I dont really understand why they work, people can set it up using keyboard override if they want, but they will only do it if it works (ie not on a mac). Revisit if crossterm fixes their issue

extrawurst commented on src/components/textinput.rs:
```diff
@@ -1,52 +1,55 @@
+	SelectionEndPending,
+}
+type TextAreaComponent = TextArea<'static>;
+#[allow(clippy::struct_excessive_bools)]
```
this allow is not needed anymore

extrawurst commented on src/components/textinput.rs:
```diff
@@ -84,14 +86,30 @@ impl TextInputComponent {
 	}
 
 	/// Get the `msg`.
+
```
consistency: no empty line between doc comment and fn

extrawurst commented on src/components/textinput.rs:
```diff
@@ -233,96 +182,44 @@ impl TextInputComponent {
-			if ends_in_nl {
-				txt.lines.push(Line::default());
-			}
+		if self.is_visible() {
```
why recreate the whole textarea ?

extrawurst commented on src/components/textinput.rs:
```diff
@@ -109,120 +127,51 @@ impl TextInputComponent {
-		self.update_count();
+		self.msg = msg.into();
+		if self.is_visible() {
+			self.create_and_show();
```
why recreate the whole textarea ?

pm100 replied:
Dont go there :-). This code is the result of > 6 months of work arm wrestling with tui-textarea, your API into textarea (which I took as a given) and rust lifetime management.

extrawurst commented on src/components/textinput.rs:
```diff
@@ -109,120 +127,51 @@ impl TextInputComponent {
-			let is_in_word = self.at_alphanumeric(index);
-			if was_in_word && !is_in_word {
-				return Some(last_pos);
+	fn create_and_show(&mut self) {
```
lets call it `create_inner_textarea`

extrawurst commented on src/components/textinput.rs:
```diff
@@ -233,96 +182,44 @@ impl TextInputComponent {
-				self.char_count
-			))
-			.alignment(Alignment::Right);
+	fn draw_newline_hint<B: Backend>(
```
<img width="868" alt="Screenshot 2024-02-15 at 21 24 22" src="https://github.com/extrawurst/gitui/assets/776816/51286099-28b7-4a13-b6ae-940eff1d09fc">

lets use the machinery that is already there to feed this new command into the cmd-bar on the bottom see `fn commands`

pm100 replied:
i missed that, will change

extrawurst reviewed (changes requested):
thanks for sticking with this. i think we are getting closer. i added a bunch of comments here and there.

pm100 commented on src/components/commit.rs:
```diff
@@ -565,7 +565,7 @@ impl Component for CommitComponent {
 
 			if let Event::Key(e) = ev {
-				if key_match(e, self.key_config.keys.enter)
+				if key_match(e, self.key_config.keys.multiline_enter)
```
Its a bit more fiddly than that . What should the behavior be for other multi-line edit boxes. For example 'reword commit'. I also wonder about stash name dialog, I changed it to be single-line but I think thats wrong, it can be a multi-line commit style message. Do all those behave the same way? If so (and I should check that they do work) them maybe 'commit' isnt the best name

pm100 commented:
OK. we are close :-)

I did everything you said. 

Notes: 

Should stash msg dialog be multi or single? I made it single but I think it should be multiple. Not sure tho. If multiple I need to change the plumbing around it too.

The defaults are Enter = commit, ctrl-r = newline. I have test both these

```
(
commit: Some((code:Char('o') , modifiers:"CONTROL")),
newline: Some((code:Enter, modifiers:"NONE")),
)
```
which is the one you asked for at the start

and

```
(
commit: Some((code:Enter , modifiers:"NONE")),
newline: Some((code:Enter, modifiers:"SHIFT")),
)
```

Which is what I originally had but doesnt work on a mac - as you found out.
